pkg/asset/agent/joiner/clusterinfo.go

```diff
@@ -0,0 +1,232 @@
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/agent"
+	"github.com/openshift/installer/pkg/asset/agent/workflow"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// ClusterInfo it's an asset used to retrieve config info
+// from an already existing cluster. A number of different resources
+// are inspected to extract the required configuration.
+type ClusterInfo struct {
+	ClusterID                     string
+	Version                       string
+	ReleaseImage                  string
+	APIDNSName                    string
+	PullSecret                    string
+	Namespace                     string
+	UserCaBundle                  string
+	Proxy                         *types.Proxy
+	Architecture                  string
+	ImageDigestSources            []types.ImageDigestSource
+	DeprecatedImageContentSources []types.ImageContentSource
+	PlatformType                  hiveext.PlatformType
+	SSHKey                        string
+}
+
+var _ asset.WritableAsset = (*ClusterInfo)(nil)
+
+// Name returns the human-friendly name of the asset.
+func (ci *ClusterInfo) Name() string {
+	return "Agent Installer ClusterInfo"
+}
+
+// Dependencies returns all of the dependencies directly needed to generate
+// the asset.
+func (*ClusterInfo) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&workflow.AgentWorkflow{},
+		&AddNodesConfig{},
+	}
+}
+
+// Generate generates the ClusterInfo.
+func (ci *ClusterInfo) Generate(dependencies asset.Parents) error {
+	agentWorkflow := &workflow.AgentWorkflow{}
+	addNodesConfig := &AddNodesConfig{}
+	dependencies.Get(agentWorkflow, addNodesConfig)
+
+	if agentWorkflow.Workflow != workflow.AgentWorkflowTypeAddNodes {
+		return nil
+	}
+
+	config, err := ci.getRestConfig(addNodesConfig.Params.Kubeconfig)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveAPIDNSName(config)
+	if err != nil {
+		return err
+	}
+
+	clientset, err := configclient.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveClusterData(clientset)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveProxy(clientset)
+	if err != nil {
+		return err
+	}
+
+	k8sclientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	err = ci.retrievePullSecret(k8sclientset)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveUserTrustBundle(k8sclientset)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveArchitecture(k8sclientset)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveInstallConfigData(k8sclientset)
+	if err != nil {
+		return err
+	}
+
+	ci.Namespace = "cluster0"
+
+	return nil
+}
+
+func (ci *ClusterInfo) getRestConfig(kubeconfig string) (*rest.Config, error) {
+	var err error
+	var config *rest.Config
+
+	if kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
```

Comment:
That's the default client configuration, to support both the case when the client is executed outside the cluster (and thus a valid kubeconfig is fully required)  and when it's run inside a cluster (and this would be, in the longer term, our preferred way of running node-joiner)